service/src/main/java/io/camunda/service/UserTaskServices.java

```diff
@@ -40,11 +58,176 @@ public UserTaskServices withAuthentication(final Authentication authentication)
+
+    final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey =
+        searchResults.items().stream()
+            .collect(
+                Collectors.groupingBy(
+                    VariableEntity::processInstanceKey,
+                    Collectors.mapping(VariableEntity::scopeKey, Collectors.toSet())));
+
+    if (firstIteration) {
+      initializeScopeKeyMap(
+          scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    } else {
+      updateScopeKeyMap(scopeKeyMapByProcessInstanceKey, currentScopeKeyMapByProcessInstanceKey);
+    }
+  }
+
+  private void initializeScopeKeyMap(
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey) {
+    for (final Map.Entry<Long, Set<Long>> entry :
+        currentScopeKeyMapByProcessInstanceKey.entrySet()) {
+      final Long processInstanceId = entry.getKey();
+      final Set<Long> validScopeKeys =
+          entry.getValue().stream()
+              .filter(scopeKey -> !scopeKey.equals(processInstanceId))
+              .collect(Collectors.toSet());
+      if (!validScopeKeys.isEmpty()) {
+        scopeKeyMapByProcessInstanceKey.put(processInstanceId, validScopeKeys);
+      } else {
+        scopeKeyMapByProcessInstanceKey.put(processInstanceId, null);
+      }
+    }
+  }
+
+  private void updateScopeKeyMap(
+      final Map<Long, Set<Long>> scopeKeyMapByProcessInstanceKey,
+      final Map<Long, Set<Long>> currentScopeKeyMapByProcessInstanceKey) {
+    scopeKeyMapByProcessInstanceKey
+        .entrySet()
+        .removeIf(
+            entry -> {
+              final Long processInstanceId = entry.getKey();
+              final Set<Long> validScopeKeys =
+                  entry.getValue() != null
+                      ? entry.getValue().stream()
+                          .filter(Objects::nonNull)
+                          .collect(Collectors.toSet())
+                      : new HashSet<>();
+              final Set<Long> currentScopeKeys =
+                  currentScopeKeyMapByProcessInstanceKey.get(processInstanceId);
+
+              if (currentScopeKeys != null) {
+                validScopeKeys.retainAll(currentScopeKeys);
+              }
+
+              if (entry.getValue() == null || validScopeKeys.isEmpty()) {
+                return currentScopeKeys == null || currentScopeKeys.isEmpty();
+              }
+
+              return validScopeKeys.isEmpty()
+                  || !currentScopeKeyMapByProcessInstanceKey.containsKey(processInstanceId);
+            });
+
+    currentScopeKeyMapByProcessInstanceKey.forEach(
+        (processInstanceId, currentScopeKeys) ->
+            scopeKeyMapByProcessInstanceKey.merge(
+                processInstanceId,
+                currentScopeKeys,
+                (oldSet, newSet) -> {
+                  oldSet.addAll(newSet);
+                  return oldSet;
+                }));
```

Comment:
❓ Why must it be merged at all? Shouldn't it be enough to remove `processInstanceKeys` and `scopeKeys` from `scopeKeyMapByProcessInstanceKey`?

Reply:
Having the following case

```
pi(1) ---> fi(2) --> var: foo==123
    |
   +------> fi(3) --> var: bar == 456
```

Now doing the following user task query:
```json
{"filter": { "variable": [ {"name": foo", "eq": 123 }, {"name": bar", "eq": 456 }] } }
```

According the logic, with the `firstIteration` it will execute a variable query with `{"name": foo", "eq": 123 }`, which will return `pi(1) ---> fi(2) --> var: foo==123`. Basically, the variable `foo`:
```json
{
  "processInstanceKey": 1,
  "scopeKey": 2,
  "name": "foo",
  "value": 123
}
```

The `scopeKeyMapByProcessInstanceKey` looks like this: `{ 1: [2] }`

In the next iteration, it will combine the variable value filter `{"name": bar", "eq": 456 }` with the process instance key `1`
```json
{"filter": { "variable": [ {"name": bar", "eq": 456 }], "processInstanceKeys": 1 } }
```

It will return
```json
pi(1) ---> fi(3) --> var: bar ==456
```

The `currentScopeKeyMapByProcessInstanceKey` looks like this: `{ 1: [3] }`

Based on this, it `removeIf()`, the `scopeKeyMapByProcessInstanceKey` will be changed to `{ 1: [] }` (because `[2]` and `[3]` do not intersect. And executing the `merge()`, it will merge the `scopeKeyMapByProcessInstanceKey` will be merged to `{ 1: [3] }`

Reply:
As a result, it will return the user task (that belongs to flownode instance `3`) which would not be expected.